Code/GraphMol/Descriptors/AtomicEnvironmentVector.cpp

```diff
@@ -0,0 +1,845 @@
+//
+//  Copyright (C) 2020 Manan Goel
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include <GraphMol/RDKitBase.h>
+#include <cmath>
+#include <Numerics/EigenSerializer/EigenSerializer.h>
+#include "AtomicEnvironmentVector.h"
+#include <Eigen/Dense>
+
+using namespace Eigen;
+
+namespace RDKit {
+namespace Descriptors {
+namespace ANI {
+
+//! Calculates the value a continuous smoothening function for a distance such
+//! that values
+// greater than the cutoff give 0
+/*!
+  \param distances A 2 dimensional array of pairwise distances
+  \param cutoff    A double value signifying cutoff distance
+
+  \return 2 dimensional array containing corresponding values computed by cutoff
+  function
+*/
+template <typename Derived>
+ArrayXXd CosineCutoff(ArrayBase<Derived> *distances, double cutoff) {
+  // Cosine cutoff function assuming all distances are less than the cutoff
+  PRECONDITION(cutoff > 0.0, "Cutoff must be greater than zero");
+  PRECONDITION(((*distances) <= cutoff).count() == distances->size(),
+               "All distances must be less than the cutoff");
+  PRECONDITION(distances != nullptr, "Array of distances is NULL");
+  return 0.5 * ((*distances) * (M_PI / cutoff)).cos() + 0.5;
+}
+
+VectorXi GenerateSpeciesVector(const ROMol &mol) {
+  // Generate atom species vector as mentioned in torchani
+  // H : 0
+  // C : 1
+  // N : 2
+  // O : 3
+  // All other atoms : -1
+  auto numAtoms = mol.getNumAtoms();
+  VectorXi species(numAtoms);
+
+  for (unsigned int i = 0; i < numAtoms; i++) {
+    auto atom = mol.getAtomWithIdx(i);
+
+    switch (atom->getAtomicNum()) {
+      case 1:
+        species[i] = 0;
+        break;
+      case 6:
+        species[i] = 1;
+        break;
+      case 7:
+        species[i] = 2;
+        break;
+      case 8:
+        species[i] = 3;
+        break;
+      default:
+        throw ValueErrorException("Atom Type Not Supported");
+    }
+  }
+
+  return species;
+}
+
+VectorXi GenerateSpeciesVector(const int *atomNums, unsigned int numAtoms) {
+  PRECONDITION(atomNums != nullptr, "Array of atom types is NULL")
+  VectorXi species(numAtoms);
+
+  for (unsigned int i = 0; i < numAtoms; i++) {
+    switch (atomNums[i]) {
```

Comment:
This duplicates the switch statement from above, it seems we should have this logic in one place.